sycl/include/CL/sycl/detail/pi.h

```diff
@@ -928,6 +929,142 @@ pi_result piEnqueueMemUnmap(
   const pi_event * event_wait_list,
   pi_event *       event);
 
+///
+// USM
+///
+typedef enum {
+  PI_USM_HOST_SUPPORT          = CL_DEVICE_HOST_MEM_CAPABILITIES_INTEL,
+  PI_USM_DEVICE_SUPPORT        = CL_DEVICE_DEVICE_MEM_CAPABILITIES_INTEL,
+  PI_USM_SINGLE_SHARED_SUPPORT = CL_DEVICE_SINGLE_DEVICE_SHARED_MEM_CAPABILITIES_INTEL,
+  PI_USM_CROSS_SHARED_SUPPORT  = CL_DEVICE_CROSS_DEVICE_SHARED_MEM_CAPABILITIES_INTEL,
+  PI_USM_SYSTEM_SHARED_SUPPORT = CL_DEVICE_SHARED_SYSTEM_MEM_CAPABILITIES_INTEL
+} _pi_usm_capability_query;
+
+typedef enum : pi_bitfield {
+  PI_USM_ACCESS                   = CL_UNIFIED_SHARED_MEMORY_ACCESS_INTEL,
+  PI_USM_ATOMIC_ACCESS            = CL_UNIFIED_SHARED_MEMORY_ATOMIC_ACCESS_INTEL,
+  PI_USM_CONCURRENT_ACCESS        = CL_UNIFIED_SHARED_MEMORY_CONCURRENT_ACCESS_INTEL,
+  PI_USM_CONCURRENT_ATOMIC_ACCESS = CL_UNIFIED_SHARED_MEMORY_CONCURRENT_ATOMIC_ACCESS_INTEL
+} _pi_usm_capabilities;
+
+typedef enum {
+  PI_MEM_ALLOC_TYPE        = CL_MEM_ALLOC_TYPE_INTEL,
+  PI_MEM_ALLOC_BASE_PTR    = CL_MEM_ALLOC_BASE_PTR_INTEL,
+  PI_MEM_ALLOC_SIZE        = CL_MEM_ALLOC_SIZE_INTEL,
+  PI_MEM_ALLOC_INFO_DEVICE = CL_MEM_ALLOC_INFO_DEVICE_INTEL,
+  PI_MEM_ALLOC_INFO_TBD0   = CL_MEM_ALLOC_INFO_TBD0_INTEL,
+  PI_MEM_ALLOC_INFO_TBD1   = CL_MEM_ALLOC_INFO_TBD1_INTEL,
+} _pi_mem_info;
+
+typedef enum {
+  PI_MEM_TYPE_UNKNOWN = CL_MEM_TYPE_UNKNOWN_INTEL,
+  PI_MEM_TYPE_HOST    = CL_MEM_TYPE_HOST_INTEL,
+  PI_MEM_TYPE_DEVICE  = CL_MEM_TYPE_DEVICE_INTEL,
+  PI_MEM_TYPE_SHARED  = CL_MEM_TYPE_SHARED_INTEL
+} _pi_usm_type;
+
+typedef enum : pi_bitfield  {
+  PI_MEM_ALLOC_FLAGS = CL_MEM_ALLOC_FLAGS_INTEL
+} _pi_usm_mem_properties;
+
+typedef enum : pi_bitfield {
+  PI_USM_MIGRATION_TBD0 = (1 << 0)
+} _pi_usm_migration_flags;
+
+typedef enum {
+  PI_INDIRECT_HOST_ACCESS   = CL_KERNEL_EXEC_INFO_INDIRECT_HOST_ACCESS_INTEL,
+  PI_INDIRECT_DEVICE_ACCESS = CL_KERNEL_EXEC_INFO_INDIRECT_DEVICE_ACCESS_INTEL,
+  PI_INDIRECT_SHARED_ACCESS = CL_KERNEL_EXEC_INFO_INDIRECT_SHARED_ACCESS_INTEL,
+  PI_USM_PTRS               = CL_KERNEL_EXEC_INFO_USM_PTRS_INTEL
+} _pi_usm_kernel_exec_info;
+
+typedef _pi_usm_capability_query  pi_usm_capability_query;
+typedef _pi_usm_capabilities      pi_usm_capabilities;
+typedef _pi_mem_info              pi_mem_info;
+typedef _pi_usm_type              pi_usm_type;
+typedef _pi_usm_mem_properties    pi_usm_mem_properties;
+typedef _pi_usm_kernel_exec_info  pi_usm_kernel_exec_info;
+typedef _pi_usm_migration_flags   pi_usm_migration_flags;
+
+pi_result piHostMemAlloc(
```

Comment:
Please provide documentation (doxygen) for each new PI interface.
Few general comment:
1) we should start these with "piext", since we don't want to request all SYCL back-ends to support it, right? (btw how to query if USM is supported?)
2) the names should probably include USM to not introduce confusion with other piMem* methods (can we maybe extend some of the existing piMem's to handle USM as well?)
3) the naming convention is to give object name first so maybe piextUSMHostAlloc?
4) the rest of PI API provide ref-count capabilities through OpenCL-like retain/release methods. why should these be different?

Reply:
1) Sure
2) Sure
3) Sure
4) USM allocations aren't reference counted, so I'm not sure what you're asking.

Reply:
dev.get_info<info::device::usm_*_allocations>() tests whether it is supported.

Reply:
What PI calls does this translate into?

Reply:
> 4\. USM allocations aren't reference counted, so I'm not sure what you're asking.

I am asking if we should make them reference counted like other objects in PI.

Reply:
> > dev.get_infoinfo::device::usm_*_allocations() tests whether it is supported.
> 
> What PI calls does this translate into?
piDeviceGetInfo

Reply:
> > 4. USM allocations aren't reference counted, so I'm not sure what you're asking.
> 
> I am asking if we should make them reference counted like other objects in PI.

No - they're raw pointers.  We can't shove a wrapper around them.  Users can use a std::shared_ptr or std::unique_ptr on the host if they want.  Won't work across host and device.